src/sync.ts

```diff
@@ -1600,6 +1601,13 @@ export class SyncApi {
         return (obj.events as Array<IStrippedState | IRoomEvent | IStateEvent | IMinimalEvent>).map(function(e) {
             if (room) {
                 e["room_id"] = room.roomId;
+                if ("unsigned" in e) {
+                    const threadRel: IThreadBundledRelationship =
+                        e.unsigned?.["m.relations"]?.[THREAD_RELATION_TYPE.name];
+                    if (threadRel) {
+                        threadRel.latest_event.room_id = room.roomId;
+                    }
+                }
```

Comment:
```suggestion
                  const threadRel: IThreadBundledRelationship =
                      e.unsigned?.["m.relations"]?.[THREAD_RELATION_TYPE.name];
                  if (threadRel) {
                      threadRel.latest_event.room_id = room.roomId;
                  }
```

You're already using `e.unsigned?.` so the if check wasn't adding anything

Reply:
So the spec is unclear on this, it only has a single example of `latest_event` and it doesn't have a room_id. Should we be doing similar trickery when we get the latest event via the /event, /context and /relations APIs? IIRC those do include a room_id but that could be a Synapse idiom

Reply:
The spec says

> The latest event should be serialised in the same form as the event itself; this includes adding any bundled aggregations for the event (and applying edits).[1](https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/3440-threading-via-relations.md#user-content-fn-1-b89185a683eddef9c6c0acdc33023cfd)

https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/3440-threading-via-relations.md

So I would argue this is a Synapse bug

Reply:
I tried moving the fix to `thread.ts` just now, but that makes my new test fail. It looks like there are other places where we process bundled relations that don't specifically name `latest_event`, so I'd need to track down all of those and apply a fix at each point. I think this *is* the most resilient solution, because it re-inserts the room ID at the earliest point possible before anything else can start trying to process the event. I'm confused about what would be wrong with this approach; worst case, this event already has a room ID and re-inserting the room ID is a no-op. If you'd like, I can add a check so that we only insert a room ID if there isn't one present already?

Reply:
And the purpose of the `if` statement is for type narrowing; ultimately it's somewhat moot because of the `any` in the record type further down the line, but it does give IDEs etc. the ability to type-check things up to `m.relations`. I can remove it if you don't think it's worth it.

Reply:
Ah, I see your point about making sure we're handling bundled events from endpoints other than `/sync` correctly. It looks like in the spec, `/sync` is the only place where `ClientEventWithoutRoomID` is used at all, FWIW.

Reply:
Yeah but equally the MSC is vague on the language, and also mutating the unsigned field returned from the server isn't something we do elsewhere to my knowledge, so would be good to stay consistent with that as it may confuse developers making use of /devtools